source/material_model/damage_rheology.cc

```diff
@@ -0,0 +1,2027 @@
+          if (crossed_transition != -1)
+            if (recrystallized_grain_size[crossed_transition] > 0.0)
+              phase_grain_size_reduction = grain_size - recrystallized_grain_size[crossed_transition];
+        }
+      else if (this->introspection().name_for_compositional_index(field_index) == "pyroxene_grain_size")
+        {
+          phase_grain_size_reduction = 0.0;
+        }
+
+      if (grain_size < 5.e-6)
+        {
+          std::cout << "Grain size is " << grain_size << "! It needs to be larger than 5e-6.\n";
+          grain_size = 5e-6;
+        }
+
+      return grain_size - original_grain_size - phase_grain_size_reduction;
+    }
+
+
+
+    template <int dim>
+    double
+    DamageRheology<dim>::
+    diffusion_viscosity (const double                  temperature,
+                         const double                  pressure,
+                         const std::vector<double>    &composition,
+                         const SymmetricTensor<2,dim> &strain_rate,
+                         const Point<dim>             &position) const
+    {
+      const SymmetricTensor<2,dim> shear_strain_rate = strain_rate - 1./dim * trace(strain_rate) * unit_symmetric_tensor<dim>();
+      const double second_strain_rate_invariant = std::sqrt(std::abs(second_invariant(shear_strain_rate)));
+
+      // TODO: make this more general, for more phases we have to average grain size somehow
+      // TODO: default when field is not given & warning
+      // limit the grain size to a global minimum
+      const std::string field_name = "olivine_grain_size";
+      const double grain_size = this->introspection().compositional_name_exists(field_name)
+                                ?
+                                composition[this->introspection().compositional_index_for_name(field_name)]
+                                :
+                                1.0;
+
+      // Currently this will never be called without adiabatic_conditions initialized, but just in case
+      const double adiabatic_pressure = this->get_adiabatic_conditions().is_initialized()
+                                        ?
+                                        this->get_adiabatic_conditions().pressure(position)
+                                        :
+                                        pressure;
+
+      // find out in which phase we are
+      const unsigned int ol_index = get_phase_index(position, temperature, adiabatic_pressure);
+
+      // TODO: we use the prefactors from Behn et al., 2009 as default values, but their laws use the strain rate
+      // and we use the second invariant --> check if the prefactors should be changed
+      double energy_term = exp((diffusion_activation_energy[ol_index] + diffusion_activation_volume[ol_index] * adiabatic_pressure)
+                               / (diffusion_creep_exponent[ol_index] * constants::gas_constant * temperature));
+      if (this->get_adiabatic_conditions().is_initialized())
```

Comment:
Can you write a comment that this is to limit the temperature dependence of viscosity?